segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+
+                if (segmentMetadata.isSealed()) {
+                    throw new StreamSegmentSealedException(streamSegmentName);
+                }
+
+                if (segmentMetadata.getOwnerEpoch() > this.epoch) {
+                    throw new StorageNotPrimaryException(streamSegmentName);
+                }
+
+                if (segmentMetadata.getLength() < offset || segmentMetadata.getStartOffset() > offset) {
+                    throw new IllegalArgumentException(String.format("offset %d is outside of valid range [%d, %d) for segment %s",
+                            offset, segmentMetadata.getStartOffset(), segmentMetadata.getLength(), streamSegmentName));
+                }
+
+                if (segmentMetadata.getStartOffset() == offset) {
+                    // Nothing to do
+                    return null;
+                }
+
+                String currentChunkName = segmentMetadata.getFirstChunk();
+                ChunkMetadata currentMetadata;
+                long oldLength = segmentMetadata.getLength();
+                long startOffset = segmentMetadata.getFirstChunkStartOffset();
+                ArrayList<String> chunksToDelete = new ArrayList<>();
+                while (currentChunkName != null) {
+                    currentMetadata = (ChunkMetadata) txn.get(currentChunkName);
+                    Preconditions.checkState(null != currentMetadata, "currentMetadata is null.");
+
+                    // If for given chunk start <= offset < end  then we have found the chunk that will be the first chunk.
+                    if ((startOffset <= offset) && (startOffset + currentMetadata.getLength() > offset)) {
+                        break;
+                    }
+
+                    startOffset += currentMetadata.getLength();
+                    chunksToDelete.add(currentMetadata.getName());
+                    segmentMetadata.decrementChunkCount();
+
+                    // move to next chunk
+                    currentChunkName = currentMetadata.getNextChunk();
+                }
+                segmentMetadata.setFirstChunk(currentChunkName);
+                segmentMetadata.setStartOffset(offset);
+                segmentMetadata.setFirstChunkStartOffset(startOffset);
+                for (String toDelete : chunksToDelete) {
+                    txn.delete(toDelete);
+                    // Adjust last chunk if required.
+                    if (toDelete.equals(segmentMetadata.getLastChunk())) {
+                        segmentMetadata.setLastChunkStartOffset(segmentMetadata.getLength());
+                        segmentMetadata.setLastChunk(null);
+                    }
+                }
+                txn.update(segmentMetadata);
+
+                // Check invariants.
+                Preconditions.checkState(segmentMetadata.getLength() == oldLength, "truncate should not change segment length");
+                segmentMetadata.checkInvariants();
+
+                // Commit system logs.
```

Comment:
Why do we need to treat system logs differently?

Reply:
The table segment for storage metadata and container metadata are special segments, they store data in storage and also used for keeping metadata. 
To avoid this circular dependency at boot time (unfortunately very tricky to handle), we write log record for any change to these segment's layout into the system log. At boot time we replay it. So that we have complete info about metadata data table.